### PR TITLE
temple ast and protocols

### PR DESCRIPTION
**Pull request by mhanberg**
[Preview](https://deploy-preview-175--mitchell-hanberg.netlify.app/temple-ast-and-protocols/)

### CODE REVIEW

File: _posts/2021-04-09-temple-ast-and-protocols.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+This brings us to our next topic, protocols!
+
+## Protocols
+
+The EEX generator step utilizes a [protocol](https://elixir-lang.org/getting-started/protocols.html) to be able to compile Temple AST into an iolist that represents EEx.
+
+Each AST module implements this protocol and this allows any protocol implementation to generate any child nodes it contains without concerning itself with the shape of the children.
+
+The impl for the `Text` node type is the easiest to understand.
```
Comment (swanson) [inline _posts/2021-04-09-temple-ast-and-protocols.md:141]: Is "impl" a common shorthand? You use "implementation" in the next paragraph

File: _posts/2021-04-09-temple-ast-and-protocols.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    end
+  end
+end
+{% endhl %}
+
+Since the implementation takes advantage of iolists, we can easily compute the final markup without maintaining any state or dealing with cumbersome return values. Once `to_eex` returns, we just run that through `:erlang.iolist_to_binary/1` and we're good to go!
+
+## What's Next
+
+With a proper AST in place, I can now move forward with the Slots API, which is the missing piece of the puzzle to make the Component API _really_ slick.
```
Comment (swanson) [inline _posts/2021-04-09-temple-ast-and-protocols.md:182]: "Named Slots API"?

**Review by swanson — COMMENTED** [review]
Looks good, the line highlighting is slick.

File: _posts/2021-04-09-temple-ast-and-protocols.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+Temple started with the ability to produce HTML at runtime, but now includes:
+
+- EEx output target
+- LiveView support (it's just EEx after all!)
+- Basic component functionality (essentially just partials)
+
+If my goals for this project are going to evolve, so does the code base. So far I've been able to accomplish this with a rather naive and imperative compilation process.
+
+I figured that writing an actual abstract syntax tree (or AST) as an intermediate format (or IF) would be the next step, and along that journey I also found a nice use case for a [protocol](https://elixir-lang.org/getting-started/protocols.html).
```
Comment (mhanberg) [inline _posts/2021-04-09-temple-ast-and-protocols.md:19]:
```suggestion
I figured that writing an actual abstract syntax tree (AST) as an intermediate format (IF) would be the next step, and along that journey I also found a nice use case for a [protocol](https://elixir-lang.org/getting-started/protocols.html).
```

File: _posts/2021-04-09-temple-ast-and-protocols.md (hunk 1 above)
Comment (mhanberg) [inline _posts/2021-04-09-temple-ast-and-protocols.md:141]:
```suggestion
The implementation for the `Text` node type is the easiest to understand.
```

File: _posts/2021-04-09-temple-ast-and-protocols.md (hunk 2 above)
Comment (mhanberg) [inline _posts/2021-04-09-temple-ast-and-protocols.md:182]:
```suggestion
With a proper AST in place, I can now move forward with the Named Slots API, which is the missing piece of the puzzle to make the Component API _really_ slick.
```

**Pull request merged**